### PR TITLE
*: implement analyze columns push down

### PR DESCRIPTION
**Pull request by lamxTyler**
PTAL @coocood  @hanfei1991  @AndreMouche

### CODE REVIEW

**Comment by coocood** [comment]
executor/new_analyze.go:150: declaration of "err" shadows declaration at executor/new_analyze.go:149
make: *** [check] Error 1

**Comment by coocood** [comment]
@lamxTyler 
We should use a constant probability to collect samples in TiKV.
If the rows count in regions are not equal, use the default sample collector will collect more samples in small regions, less samples in large regions.

**Comment by lamxTyler** [comment]
However, we may not collect enough stats if we use constant probability because we do not know the total row count. @coocood

**Comment by coocood** [comment]
@lamxTyler 
If the table is very small, we can analyze a second time with a larger probability.

**Comment by lamxTyler** [comment]
The table may change during the analyze. You cannot guarantee a larger probability gives larger samples. @coocood

**Comment by coocood** [comment]
LGTM

File: executor/new_analyze.go [hunk 1]
```diff
@@ -100,3 +103,117 @@ func (e *AnalyzeIndexExec) buildHistogram() (hist *statistics.Histogram, err err
...
+		hists = append(hists, pkHist)
+	}
+	for i, col := range e.colsInfo {
+		for j, s := range collectors[i].Samples {
+			collectors[i].Samples[j], err = tablecodec.DecodeColumnValue(s.GetBytes(), &col.FieldType, timeZone)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		hg, err := statistics.BuildColumn(e.ctx, maxBucketSize, col.ID, collectors[i].Sketch.NDV(), collectors[i].Count, collectors[i].NullCount, collectors[i].Samples)
```
Comment (hanfei1991) [inline executor/new_analyze.go]: How about only pass a collector to control the arguments count ?

File: executor/new_analyze.go [hunk 2]
```diff
@@ -100,3 +103,117 @@ func (e *AnalyzeIndexExec) buildHistogram() (hist *statistics.Histogram, err err
...
+	e.result.Fetch(e.ctx.GoCtx())
+	return nil
+}
+
+func (e *AnalyzeColumnsExec) buildHistograms() (hists []*statistics.Histogram, err error) {
+	if err = e.open(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer func() {
+		if err1 := e.result.Close(); err1 != nil {
```
[the remark this answers is not in the document]
Comment (lamxTyler) [reply executor/new_analyze.go]: Nope, this function will be called just before return, so if there is an error, the return value will be modified.

**Pull request merged**